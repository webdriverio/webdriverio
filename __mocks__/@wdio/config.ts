import { vi } from 'vitest'
import type { Options } from '../../packages/wdio-types'

import { DEFAULT_CONFIGS as DEFAULT_CONFIGS_IMPORT } from '../../packages/wdio-config/src/constants.js'
import {
    isCloudCapability as isCloudCapabilityMock,
    validateConfig as validateConfigMock
} from '../../packages/wdio-config/src/utils.js'

class ConfigParserMock {
    addService = vi.fn()
    initialize = vi.fn()
    addConfigFile = vi.fn()
    merge = vi.fn()
    autoCompile = vi.fn()

    getConfig = vi.fn().mockReturnValue({
        runnerEnv: {},
        shard: { current: 1, total: 1 },
        runner: 'local',
        outputDir: './tempDir'
    })

    getCapabilities = vi.fn().mockReturnValue([{
        browserName: 'chrome',
        specs: ['./tests/test2.js']
    }, {
        browserName: 'firefox'
    }])

    getSpecs = vi.fn().mockReturnValue(['./tests/test1.js'])
}

export const ConfigParser = ConfigParserMock
export const DEFAULT_CONFIGS = DEFAULT_CONFIGS_IMPORT as () => Omit<Options.Testrunner, 'capabilities'>
export const isCloudCapability = vi.fn().mockImplementation(isCloudCapabilityMock)
export const validateConfig = vi.fn().mockImplementation((defaults, config) => {
    const returnVal = validateConfigMock(defaults, config)
    /**
     * in order to ensure that WebdriverIO unit tests default to use the webdriver
     * package we modify the default value here
     */
    if (!returnVal.hostname) {
        returnVal.hostname = 'localhost'
    }

    return returnVal
})
