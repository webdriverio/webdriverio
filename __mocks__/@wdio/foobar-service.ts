export default class FoobarService {
    public foo = 'foobar'
}
