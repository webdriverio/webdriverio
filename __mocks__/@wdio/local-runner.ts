import { vi } from 'vitest'
import type WorkerInstance from '../../packages/wdio-local-runner/src/worker'

export default class LocalRunnerMock {
    configFile: string
    config: WebdriverIO.Config
    workerPool: Record<string, WorkerInstance> = {}
    initialize = vi.fn()

    constructor (configFile: string, config: WebdriverIO.Config) {
        this.configFile = configFile
        this.config = config
    }

    run ({ command, args, ...options }: any) {
        this.workerPool[options.cid as string] = { postMessage: vi.fn() } as unknown as WorkerInstance
        this.workerPool[options.cid].postMessage(command, args)
        return this.workerPool[options.cid]
    }
}
