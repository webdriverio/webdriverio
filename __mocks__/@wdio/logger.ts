import { vi } from 'vitest'

export const logMock = {
    error: vi.fn(),
    debug: vi.fn(),
    warn: vi.fn(),
    info: vi.fn(),
    trace: vi.fn(),
    progress: vi.fn()
}
const mock = () => logMock
mock.setLevel = vi.fn()
mock.setLogLevelsConfig = vi.fn()
mock.waitForBuffer = vi.fn()
mock.clearLogger = vi.fn()
export default mock
