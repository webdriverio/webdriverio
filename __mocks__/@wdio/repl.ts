import { vi } from 'vitest'
import repl from 'node:repl'

import { ReplConfig } from '../../packages/wdio-repl/src'

export default class WDIORepl {
    static introMessage = 'some intro from mock'
    private _config: ReplConfig
    private _isCommandRunning = false
    private _replServer: repl.REPLServer | undefined
    private _evalFn: (...args: [any]) => any
    private _startFn: (...args: [any]) => any

    constructor(config: ReplConfig) {
        this._config = Object.assign({
            commandTimeout: 5000,
            eval: this.eval.bind(this),
            prompt: '\u203A ',
            useGlobal: true,
            useColor: true
        }, config)

        this._isCommandRunning = false
        this._evalFn = vi.fn()
        this._startFn = vi.fn().mockImplementation(
            () => new Promise(
                (resolve) => setTimeout(() => resolve(this), 100)))
    }

    eval(...args: [any]) {
        this._evalFn(...args)
    }

    start(...args: [any]) {
        return this._startFn(...args)
    }
}
