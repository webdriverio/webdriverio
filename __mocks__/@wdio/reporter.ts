import { vi } from 'vitest'

import { EventEmitter } from 'node:events'
import HookStats from '../../packages/wdio-reporter/src/stats/hook.ts'
import RunnerStats from '../../packages/wdio-reporter/src/stats/runner.ts'
import SuiteStats from '../../packages/wdio-reporter/src/stats/suite.ts'
import TestStats from '../../packages/wdio-reporter/src/stats/test.ts'
import { Chalk } from '../chalk.ts'

export default class WDIOReporter extends EventEmitter {
    outputStream: { write: Function }
    failures: number
    suites: Record<string, SuiteStats>
    hooks: Record<string, HookStats>
    tests: Record<string, TestStats>
    currentSuites: SuiteStats[]
    counts: {
        suites: number
        tests: number
        hooks: number
        passes: number
        skipping: number
        failures: number
    }
    retries: number
    _chalk: Chalk
    runnerStat?: RunnerStats
    constructor (public options: any) {
        super()
        this.options = options
        this.outputStream = { write: vi.fn() }
        this.failures = 0
        this.suites = {}
        this.hooks = {}
        this.tests = {}
        this.currentSuites = []
        this.counts = {
            suites: 0,
            tests: 0,
            hooks: 0,
            passes: 0,
            skipping: 0,
            failures: 0
        }
        this.retries = 0
        this._chalk = new Chalk(!options.color ? { level : 0 } : {})
    }

    get isSynchronised () {
        return true
    }

    write (content: any) {
        this.outputStream.write(content)
    }

    /* istanbul ignore next */
    onRunnerStart () {}
    /* istanbul ignore next */
    onBeforeCommand () {}
    /* istanbul ignore next */
    onAfterCommand () {}
    /* istanbul ignore next */
    onSuiteStart () {}
    /* istanbul ignore next */
    onHookStart () {}
    /* istanbul ignore next */
    onHookEnd () {}
    /* istanbul ignore next */
    onTestStart () {}
    /* istanbul ignore next */
    onTestPass () {}
    /* istanbul ignore next */
    onTestFail () {}
    /* istanbul ignore next */
    onTestSkip () {}
    /* istanbul ignore next */
    onTestEnd () {}
    /* istanbul ignore next */
    onSuiteEnd () {}
    /* istanbul ignore next */
    onRunnerEnd () {}
}

export { HookStats, RunnerStats, SuiteStats, TestStats }
