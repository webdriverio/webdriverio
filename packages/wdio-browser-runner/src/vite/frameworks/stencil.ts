import path from 'node:path'
import url from 'node:url'

import { findStaticImports, parseStaticImport, type ParsedStaticImport } from 'mlly'
import type { InlineConfig, Plugin } from 'vite'

import { hasFileByExtensions } from '../utils.js'

const __dirname = url.fileURLToPath(new URL('.', import.meta.url))
const STENCIL_IMPORT = '@stencil/core'

interface CompilerOptions {
    baseUrl?: string
    paths?: Record<string, string[]>
    target?: string
}

export async function isUsingStencilJS(rootDir: string, options: WebdriverIO.BrowserRunnerOptions) {
    return Boolean(options.preset === 'stencil' || await hasFileByExtensions(path.join(rootDir, 'stencil.config')))
}

export async function optimizeForStencil(rootDir: string) {
    const stencilConfig = await importStencilConfig(rootDir)
    const stencilPlugins = stencilConfig.config.plugins
    const stencilOptimizations: InlineConfig = {
        plugins: [await stencilVitePlugin(rootDir)],
        optimizeDeps: { include: [] as string[] }
    }

    if (stencilPlugins) {
        const esbuildPlugin = stencilPlugins.find((plugin: any) => plugin.name === 'esbuild-plugin')
        if (esbuildPlugin) {
            stencilOptimizations.optimizeDeps?.include?.push(...esbuildPlugin.options.include)
        }
    }

    /**
     * testing helper from the stencil core package is unfortunately exported as CJS
     * module, in order to be able to use it in the browser we have to optimize it
     * it to compile it to ESM
     */
    stencilOptimizations.optimizeDeps?.include?.push(
        '@wdio/browser-runner/stencil > @stencil/core/internal/testing/index.js'
    )
    return stencilOptimizations
}

async function stencilVitePlugin(rootDir: string): Promise<Plugin> {
    const { transpileSync, ts } = await import('@stencil/core/compiler/stencil.js')
    const stencilHelperPath = path.resolve(__dirname, '..', '..', 'browser', 'integrations', 'stencil.js')
    return {
        name: 'wdio-stencil',
        enforce: 'pre',
        resolveId (source) {
            if (source === '@wdio/browser-runner/stencil') {
                return stencilHelperPath
            }
        },
        transform: function (code, id) {
            const staticImports = findStaticImports(code)
            const stencilImports = staticImports
                .filter((imp) => imp.specifier === STENCIL_IMPORT)
                .map((imp) => parseStaticImport(imp))
            const isStencilComponent = stencilImports.some((imp) => 'Component' in (imp.namedImports || {}))

            /**
             * if file doesn't define a Stencil component
             */
            if (!isStencilComponent) {
                /**
                 * if a test imports the `@wdio/browser-runner/stencil` package we want to automatically
                 * import `h` and `Fragment` from the `@stencil/core` package
                 */
                const stencilHelperImport = staticImports.find((imp) => imp.specifier === '@wdio/browser-runner/stencil')
                if (stencilHelperImport) {
                    const imports = parseStaticImport(stencilHelperImport)
                    if ('render' in (imports.namedImports || {})) {
                        code = injectStencilImports(code, stencilImports)
                    }
                }
                return { code }
            }

            const tsCompilerOptions = getCompilerOptions(ts, rootDir)
            const opts = {
                componentExport: 'module',
                componentMetadata: 'compilerstatic',
                coreImportPath: '@stencil/core/internal/client',
                currentDirectory: rootDir,
                file: path.basename(id),
                module: 'esm',
                sourceMap: 'inline',
                style: 'static',
                proxy: 'defineproperty',
                styleImportData: 'queryparams',
                transformAliasedImportPaths: process.env.__STENCIL_TRANSPILE_PATHS__ === 'true',
                target: tsCompilerOptions?.target || 'es2018',
                paths: tsCompilerOptions?.paths,
                baseUrl: tsCompilerOptions?.baseUrl,
            }

            const transpiledCode = transpileSync(code, opts)

            /**
             * StencilJS applies only a getter to the component without having a setter defined.
             * This causes issue in the browser as there is a check that the setter is defined
             * if the getter is defined. We can work around this by defining a setter.
             */
            let transformedCode = transpiledCode.code.replace(
                'static get style()',
                'static set style(_) {}\n    static get style()'
            )

            /**
             * StencilJS does not import the `h` or `Fragment` function by default. We need to add it so the user
             * doesn't need to.
             */
            transformedCode = injectStencilImports(transformedCode, stencilImports)

            /**
             * Ensure that CSS imports by Stencil have an `&inline` query parameter
             */
            findStaticImports(transformedCode)
                .filter((imp) => imp.specifier.includes('&encapsulation=shadow'))
                .forEach((imp) => {
                    const cssPath = path.resolve(path.dirname(id), imp.specifier)
                    transformedCode = transformedCode.replace(
                        imp.code,
                        `import ${imp.imports.trim()} from '/@fs/${cssPath}&inline';\n`
                    )
                })

            return {
                ...transpiledCode,
                code: transformedCode,
                inputFilePath: id
            }
        }
    }
}

/**
 * StencilJS does not import the `h` or `Fragment` function by default. We need to add it so the user
 * doesn't need to.
 */
function injectStencilImports(code: string, imports: ParsedStaticImport[]) {
    const hasRenderFunctionImport = imports.some((imp) => 'h' in (imp.namedImports || {}))
    if (!hasRenderFunctionImport) {
        code = `import { h } from '@stencil/core/internal/client';\n${code}`
    }
    const hasFragmentImport = imports.some((imp) => 'Fragment' in (imp.namedImports || {}))
    if (!hasFragmentImport) {
        code = `import { Fragment } from '@stencil/core/internal/client';\n${code}`
    }

    return code
}

let _tsCompilerOptions: CompilerOptions | null = null

/**
 * Read the TypeScript compiler configuration file from disk
 * @param rootDir the location to search for the config file
 * @returns the configuration, or `null` if the file cannot be found
 */
function getCompilerOptions(ts: any, rootDir: string): CompilerOptions | null {
    if (_tsCompilerOptions) {
        return _tsCompilerOptions
    }

    if (typeof rootDir !== 'string') {
        return null
    }

    const tsconfigFilePath = ts.findConfigFile(rootDir, ts.sys.fileExists)
    if (!tsconfigFilePath) {
        return null
    }

    const tsconfigResults = ts.readConfigFile(tsconfigFilePath, ts.sys.readFile)

    if (tsconfigResults.error) {
        throw new Error(tsconfigResults.error)
    }

    const parseResult = ts.parseJsonConfigFileContent(
        tsconfigResults.config,
        ts.sys,
        rootDir,
        undefined,
        tsconfigFilePath,
    )

    _tsCompilerOptions = parseResult.options
    return _tsCompilerOptions
}

/**
 * helper method to import a Stencil config file
 */
export async function importStencilConfig(rootDir: string) {
    const configPath = path.join(rootDir, 'stencil.config.ts')
    const config = await import(configPath).catch(() => ({ config: {} }))

    /**
     * if we import the config within a CJS environment we need to
     * access the default property even though there is a named export
     */
    if ('default' in config) {
        return config.default
    }

    return config
}
