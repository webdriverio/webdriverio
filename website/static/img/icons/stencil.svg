<?xml version="1.0" encoding="UTF-8"?>
<svg width="256px" height="197px" viewBox="0 0 256 197" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" preserveAspectRatio="xMidYMid">
    <title>Stencil</title>
    <g>
        <path d="M193.065354,138.495015 L139.731692,196.923077 L41.5985231,196.923077 L94.4010831,138.495015 L193.065354,138.495015 Z M256,69.2476062 L202.694892,127.675077 L0,127.675077 L53.3053046,69.2476062 L256,69.2476062 Z M214.399015,2.13162821e-14 L161.311508,58.4277662 L62.9326769,58.4277662 L116.098954,2.13162821e-14 L214.399015,2.13162821e-14 Z" fill="#000000"></path>
    </g>
</svg>
