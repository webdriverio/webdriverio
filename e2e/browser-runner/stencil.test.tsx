import { $, browser, expect } from '@wdio/globals'
import { render, waitForChanges } from '@wdio/browser-runner/stencil'

import { AppProfile } from './components/StencilComponent.jsx'
import { NestedComponent } from './components/StencilComponentNested.jsx'
import { NoShadowComponent } from './components/StencilComponentNoShadow.jsx'

describe('Stencil Component Testing', () => {
    it('should render component correctly', async () => {
        const page = render({
            components: [AppProfile, NestedComponent],
            autoApplyChanges: true,
            template: () => (
                // @ts-ignore: types don't exist as we don't compile the components with Stencil
                <app-profile match={{ params: { name: 'stencil' } }}></app-profile>
            )
        })

        expect(page.container.tagName.toLowerCase()).toBe('stencil-stage')
        expect(page.root.tagName.toLowerCase()).toBe('app-profile')
        expect(typeof (await page.$container).elementId).toBe('string')
        expect(typeof (await page.$root).elementId).toBe('string')

        await expect($('>>>.app-profile')).toHaveText(
            expect.stringContaining('Hello! My name is Stencil.')
        )

        /**
         * this assertion for Safari due to: https://github.com/w3c/webdriver/issues/1786
         */
        if ((browser.capabilities as WebdriverIO.Capabilities).browserName?.toLowerCase() !== 'safari') {
            await expect($('>>>.app-profile')).toHaveText(
                expect.stringContaining('I am a nested component!')
            )
        }

        await expect($('>>>.app-profile').getCSSProperty('font-weight'))
            .toMatchInlineSnapshot(`
          {
            "parsed": {
              "string": "700",
              "type": "number",
              "unit": "",
              "value": 700,
            },
            "property": "font-weight",
            "value": 700,
          }
        `)
    })

    it('can determine root if rendered somewhere nested', async () => {
        const page = render({
            components: [AppProfile, NestedComponent],
            template: () => (
                <div>
                    <div>
                        {/* @ts-ignore: types don't exist as we don't compile the components with Stencil */}
                        <app-profile match={{ params: { name: 'stencil' } }}></app-profile>
                    </div>
                </div>
            )
        })

        expect(page.root.tagName.toLowerCase()).toBe('app-profile')
        expect(page.root.parentElement?.parentElement?.parentElement?.tagName.toLowerCase()).toBe('stencil-stage')
    })

    it('can render via html', async () => {
        const page = render({
            components: [NestedComponent],
            html: '<nested-component></nested-component>'
        })

        await expect(page.$root.$('>>> i')).toHaveText('I am a unknown!')
    })

    it('can wait for changes', async () => {
        const page = render({
            components: [NoShadowComponent],
            html: '<no-shadow-component></no-shadow-component>'
        })

        expect(page.root.outerHTML).toBe('<no-shadow-component></no-shadow-component>')
        await waitForChanges()
        expect(page.root.outerHTML).toBe('<no-shadow-component>Hello World!</no-shadow-component>')
    })

    it('can unmount', async () => {
        const page = render({
            components: [NestedComponent],
            html: '<nested-component></nested-component>'
        })

        await expect(page.root).toBeExisting()
        page.unmount()
        await expect(page.root).not.toBeExisting()
    })

    it('can auto peirce shadow dom', async () => {
        if ((browser.capabilities as WebdriverIO.Capabilities).browserName?.toLowerCase() === 'safari') {
            return
        }
        render({
            components: [AppProfile, NestedComponent],
            template: () => (
                <div>
                    {/* @ts-ignore: types don't exist as we don't compile the components with Stencil */}
                    <nested-component id="first transparent component"></nested-component>
                    <div className="nested">
                        {/* @ts-ignore: types don't exist as we don't compile the components with Stencil */}
                        <nested-component id="transparent component in a nested context"></nested-component>
                    </div>
                    <div className="nested second">
                        {/* @ts-ignore: types don't exist as we don't compile the components with Stencil */}
                        <nested-component id="transparent component in a second nested context"></nested-component>
                        {/* @ts-ignore: types don't exist as we don't compile the components with Stencil */}
                        <app-profile match={{ params: { name: 'stencil' } }}></app-profile>
                    </div>
                </div>
            )
        })

        await expect($('i')).toHaveText('I am a first transparent component!')
        await expect($('.nested i')).not.toBeExisting()

        /**
         * is able to find element within multiple nested levels of shadow dom
         * .second > app-profile#shadow-root > nested-component#shadow-root > i
         */
        await expect($('.second').$('app-profile').$('i'))
            .toHaveText('I am a nested component!')
        /**
         * finds first <i/> in the first nested component element
         */
        await expect($('.nested').$('i'))
            .toHaveText('I am a transparent component in a nested context!')

        /**
         * can combine different selector strategies
         */
        await expect($('.nested').$('=I am a link')).toBePresent()
        await expect($('.nested').$('*=a link')).toBePresent()

        expect(await $$('i').map((el) => el.getText())).toEqual([
            'I am a first transparent component!',
            'I am a transparent component in a nested context!',
            'I am a transparent component in a second nested context!',
            'I am a nested component!'
        ])
        expect(await $('.nested').$$('i').map((el) => el.getText())).toEqual([
            'I am a transparent component in a nested context!',
        ])
        expect(await $('.second').$$('i').map((el) => el.getText())).toEqual([
            'I am a transparent component in a second nested context!',
            'I am a nested component!'
        ])

        await expect($('nested-component')).toMatchSnapshot()
        await expect($('.nested')).toMatchSnapshot()
        await expect($('.second')).toMatchSnapshot()
    })
})
