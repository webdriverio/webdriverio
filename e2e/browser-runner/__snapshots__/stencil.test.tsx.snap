// Snapshot v1

exports[`Stencil Component Testing > can auto peirce shadow dom 1`] = `
"<nested-component>
  <template shadowroot="open" shadowrootmode="open">
    <i>I am a first transparent component!</i>
    <a href="#">I am a link</a>
  </template>
</nested-component>"
`;

exports[`Stencil Component Testing > can auto peirce shadow dom 2`] = `
"<div class="nested">
  <nested-component>
    <template shadowroot="open" shadowrootmode="open">
      <i>I am a transparent component in a nested context!</i>
      <a href="#">I am a link</a>
    </template>
  </nested-component>
</div>"
`;

exports[`Stencil Component Testing > can auto peirce shadow dom 3`] = `
"<div class="nested second">
  <nested-component>
    <template shadowroot="open" shadowrootmode="open">
      <i>I am a transparent component in a second nested context!</i>
      <a href="#">I am a link</a>
    </template>
  </nested-component>
  <app-profile>
    <template shadowroot="open" shadowrootmode="open">
      <style>.app-profile { font-weight: bold; }</style>
      <div class="app-profile">
        <p>Hello! My name is Stencil.</p>
        <nested-component>
          <template shadowroot="open" shadowrootmode="open">
            <i>I am a nested component!</i>
            <a href="#">I am a link</a>
          </template>
        </nested-component>
      </div>
    </template>
  </app-profile>
</div>"
`;
